File: bench/nested.js
// The JavaScript twin of nested.sl: loops.js's loop inside a function that also holds a closure over
// one of its names.

function run(pairs) {
    let total = 0;
    let turns = 0;
    const scale = 1;
    const weigh = (v) => v * scale;

    while (turns < 6000) {
        for (const [a, b] of pairs) {
            total = total + weigh(a * b);
        }

        turns = turns + 1;
    }

    return total;
}

const pairs = [];

for (let i = 0; i < 1000; i++) {
    pairs.push([i, i + 1]);
}

console.log(run(pairs));

File: bench/nested.rb
# The Ruby twin of nested.sl: loops.rb's loop inside a method that also holds a closure over one of
# its names. `each` with a two-parameter block unpacks each pair in the block's head, which is what
# slate's `for [a, b] in pairs` does -- and it is a block call per element, which is Ruby's walk.

def run(pairs)
  total = 0
  turns = 0
  scale = 1
  weigh = ->(v) { v * scale }

  while turns < 6000
    pairs.each do |a, b|
      total = total + weigh.(a * b)
    end

    turns = turns + 1
  end

  total
end

puts run((0...1000).map { |i| [i, i + 1] })
